smarty/forcefield.py

```diff
@@ -1427,25 +1523,63 @@ def createForce(self, system, topology, verbose=False, **kwargs):
+        impropers = ImproperDict()
+        for improper in self._impropertorsiontypes:
+            for atom_indices in topology.unrollSMIRKSMatches(improper.smirks, aromaticity_model = self.ff._aromaticity_model):
+                impropers[atom_indices] = improper
+
 
         if verbose:
             print('')
-            print('PeriodicTorsionGenerator:')
+            print('PeriodicTorsionGenerator Propers:')
             print('')
-            for torsion in self._torsiontypes:
+            for torsion in self._propertorsiontypes:
                 print('%64s : %8d matches' % (torsion.smirks, len(topology.unrollSMIRKSMatches(torsion.smirks, aromaticity_model = self.ff._aromaticity_model))))
             print('')
+            print('PeriodicTorsionGenerator Impropers:')
+            print('')
+            for improper in self._impropertorsiontypes:
+                print('%64s : %8d matches' % (improper.smirks, len(topology.unrollSMIRKSMatches(improper.smirks, aromaticity_model = self.ff._aromaticity_model))))
+            print('')
 
-        # Add all torsions to the system.
+        # Add all proper torsions to the system.
         for (atom_indices, torsion) in torsions.items():
+            # Ensure atoms are actually bonded correct pattern in Topology
+            assert topology._isBonded(atom_indices[0], atom_indices[1]), 'Atom indices %d and %d are not bonded in topology' % (atom_indices[0], atom_indices[1])
+            assert topology._isBonded(atom_indices[1], atom_indices[2]), 'Atom indices %d and %d are not bonded in topology' % (atom_indices[1], atom_indices[2])
+            assert topology._isBonded(atom_indices[2], atom_indices[3]), 'Atom indices %d and %d are not bonded in topology' % (atom_indices[2], atom_indices[3])
+
             for (periodicity, phase, k) in zip(torsion.periodicity, torsion.phase, torsion.k):
                 force.addTorsion(atom_indices[0], atom_indices[1], atom_indices[2], atom_indices[3], periodicity, phase, k)
-
         if verbose: print('%d torsions added' % (len(torsions)))
 
+
+        # Add all improper torsions to the system
+        for (atom_indices, improper) in impropers.items():
+            # Ensure atoms are actually bonded correct pattern in Topology
+            # For impropers, central atom is atom 1
+            assert topology._isBonded(atom_indices[0], atom_indices[1]), 'Atom indices %d and %d are not bonded in topology' % (atom_indices[0], atom_indices[1])
+            assert topology._isBonded(atom_indices[1], atom_indices[2]), 'Atom indices %d and %d are not bonded in topology' % (atom_indices[1], atom_indices[2])
+            assert topology._isBonded(atom_indices[1], atom_indices[3]), 'Atom indices %d and %d are not bonded in topology' % (atom_indices[1], atom_indices[3])
+
+            # Impropers are applied to all six paths around the trefoil
+            for (periodicity, phase, k) in zip(improper.periodicity, improper.phase, improper.k):
```

Comment:
You can use [`permutations`](https://docs.python.org/2/library/itertools.html#itertools.permutations) from [`itertools`](https://docs.python.org/2/library/itertools.html) to avoid having to manually expand all permutations